gitlab/utils.py

```diff
@@ -56,25 +56,33 @@ def copy_dict(dest: Dict[str, Any], src: Dict[str, Any]) -> None:
             dest[k] = v
 
 
-def _url_encode(id: str) -> str:
-    """Encode/quote the characters in the string so that they can be used in a path.
+class EncodedId(str):
+    """A custom `str` class that will return the URL-encoded value of the string.
 
-    Reference to documentation on why this is necessary.
-
-    https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding
-
-    If using namespaced API requests, make sure that the NAMESPACE/PROJECT_PATH is
-    URL-encoded. For example, / is represented by %2F
+      * Using it recursively will only url-encode the value once.
+      * Can accept either `str` or `int` as input value.
+      * Can be used in an f-string and output the URL-encoded string.
 
-    https://docs.gitlab.com/ee/api/index.html#path-parameters
+    Reference to documentation on why this is necessary.
 
-    Path parameters that are required to be URL-encoded must be followed. If not, it
-    doesn’t match an API endpoint and responds with a 404. If there’s something in front
-    of the API (for example, Apache), ensure that it doesn’t decode the URL-encoded path
-    parameters.
+    See::
 
+        https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding
+        https://docs.gitlab.com/ee/api/index.html#path-parameters
     """
-    return urllib.parse.quote(id, safe="")
+
+    # mypy complains if return type other than the class type. So we ignore issue.
+    def __new__(  # type: ignore
+        cls, value: Union[str, int, "EncodedId"]
+    ) -> Union[int, "EncodedId"]:
+        if isinstance(value, (int, EncodedId)):
+            return value
+
+        if not isinstance(value, str):
+            raise TypeError(f"Unsupported type received: {type(value)}")
+        # Set the value our string will return
```

Comment:
I feel like without the `__init__` this is now extremely obvious and we don't need the comment, WDYT?
```suggestion

```